inginious/frontend/plugins/task_hints/pages/user_hint_manager.py

```diff
@@ -0,0 +1,134 @@
+            }
+            if hints[key]["id"] in [hint["id"] for i, hint in enumerate(allowed_hints)]:
+
+                parse_content = self.parse_rst_content(hints[key]["content"])
+                to_show_hint_content["content"] = parse_content
+                to_show_hint_content["allowed_to_see"] = True
+                to_show_hint_content["penalty"] = [hint["penalty"] for i, hint in enumerate(allowed_hints) if hint["id"] == hints[key]["id"]]
+
+            to_show_hints[key] = to_show_hint_content
+
+        return to_show_hints
+
+    def update_allowed_hints_list(self, user_hints, hints):
+
+        """ Method to compare the user hints with the task hints, and update them"""
+        allowedHints = user_hints["allowedHints"]
+        hints_id = {key:hint["id"] for key, hint in hints.items()}
+        for hint in allowedHints:
+            if not hint["id"] in hints_id.values():
+                self.delete_hint_from_allowed(hint)
+                a = 1
+
+        self.update_total_penalty()
+
+    def create_hint_list_for_user(self):
+
+        user_hints = self._database.user_hints.insert({"taskid": self._task_id, "username": self._username, "allowedHints": [], "penalty": 0})
+        return user_hints
+
+    def check_allowed_hint_in_database(self, hint_id):
+
+        """ Method to check if the hint is already in the user hints """
+        user_hints = self._database.user_hints.find_one({"taskid": self._task_id, "username": self._username})
+        allowed_hints = user_hints["allowedHints"]
+
+        if hint_id in [allowed_hints[i]["id"] for i, value in enumerate(allowed_hints)]:
+            return True
+
+        return False
+
+    def delete_hint_from_allowed(self, hint_id):
+
+        """ Method to delete a hint from the user allowed hints"""
+        self._database.user_hints.find_one_and_update(
+            {"taskid": self._task_id, "username": self._username},
+            {"$pull": {
+                "allowedHints": hint_id
+            }
+        })
+
+    def add_hint_on_allowed(self, hint_id, task_hints):
+
+        """ Method to add the new unlocked hint in the user allowed hints """
+        if not self.check_allowed_hint_in_database(hint_id):
+
+            self._database.user_hints.find_one_and_update({"taskid": self._task_id, "username": self._username},
+                                                                { "$push": {
+                                                                    "allowedHints": {
+                                                                        "penalty": task_hints[hint_id]["penalty"],
+                                                                        "id": task_hints[hint_id]["id"]
+                                                                        }
+                                                                    }
+                                                                })
+            total_penalty = self.update_total_penalty()
```

Comment:
Rather than calling this method every time, why don't you just sum the penalty of the added hint to the total penalty?